hal/rfcs/0000-spi-overhaul.md

```diff
@@ -0,0 +1,216 @@
+| - | introduce `void *ctx` as an argument of `spi_transfer_async` | This argument shall be passed to the invocation of the event handler.
+| `void spi_transfer_async(...)` | `bool spi_transfer_async(...)` | `spi_transfer_async(...)` now returns a true if the transfer was scheduled and false otherwise.</br>E.g.: if the peripheral is already busy with another transfer.
+| `spi_transfer_async(..., uint32_t event, ...)` | argument removed | the callback will now be invoked on any event with the event as an argument.
+| `spi_busy(spi)` | removed | This method is never used outside of some of the hal implementations.
+| `spi_master_write(...)` `spi_slave_write(...)` `spi_slave_read(...)` `spi_slave_receive(...)` | removed | Writing data symbol by symbol is very inefficient and should not be encouraged. It can still be achieved by passing a pointer to the symbol to `spi_transfer`.
+| `spi_active(spit_t *obj)` | removed | as the async call back is now always invoked on async operation termination (unless cancelled), this status can be tracked from driver layer without any hal request.
+| `uint32_t spi_irq_handler_asynch(spi_t *obj)` | removed | as the event is now passed as an argument to the callback this method is no longer required.
+
+### The new API
+
+```c
+typedef struct {
+    /** Minimum frequency supported must be set by target device and it will be assessed during
+     *  testing.
+     */
+    uint32_t    minimum_frequency;
+    /** Maximum frequency supported must be set by target device and it will be assessed during
+     *  testing.
+     */
+    uint32_t    maximum_frequency;
+    /** Each bit represents the corresponding word length. lsb => 1bit, msb => 32bit. */
+    uint32_t    word_length;
+    bool        support_slave_mode; /**< If true, the device can handle SPI slave mode. */
+} spi_capabilities_t;
+
+typedef enum _spi_mode_t {
+  SPI_MODE_IDLE_LOW_SAMPLE_FIRST_EDGE,
+  SPI_MODE_IDLE_LOW_SAMPLE_SECOND_EDGE,
+  SPI_MODE_IDLE_HIGH_SAMPLE_FIRST_EDGE,
+  SPI_MODE_IDLE_HIGH_SAMPLE_SECOND_EDGE,
+} spi_mode_t;
+
+typedef enum _spi_bit_ordering_t {
+  SPI_BIT_ORDERING_MSB_FIRST,
+  SPI_BIT_ORDERING_LSB_FIRST,
+} spi_bit_ordering_t;
+
+typedef void (*spi_async_handler_f)(spi_t *obj, void *ctx, spi_async_event_t event);
+
+SPIName spi_get_module(PINName MISO, PINName MOSI, PINName MCLK);
+void spi_get_capabilities(SPIName name, PINName SS, spi_capabilities_t *cap);
+
+void spi_init(spi_t *obj, bool is_slave, PINName MISO, PINName MOSI, PINName MCLK, PINName SS);
+void spi_format(spi_t *obj, uint8_t bits, spi_mode_t mode, spi_bit_ordering_t bit_ordering);
+uint32_t spi_frequency(spi_t *obj, uint32_t hz);
+void spi_transfer(spi_t *obj, const void *tx, uint32_t tx_len, void *rx, uint32_t rx_len, const void *fill_symbol);
+bool spi_transfer_async(spi_t *obj, const void *tx, uint32_t tx_len, void *rx, uint32_t rx_len, const void *fill_symbol, spi_async_handler_f handler, void *ctx, DMAUsage hint);
```

Comment:
`DMAUsage` can be removed, or we keep it and have DMA later?

Reply:
AFAIK the story around DMA is not very clear so it was left here to avoid introducing unexpected breaking change.